### ISSUE TITLE
Back propagation math/training

### ISSUE THREAD

**Issue by GaelGil** [GaelGil/notebooks#14]
![back_propogation](https://user-images.githubusercontent.com/31393159/141030033-474cfd79-3bd9-47c5-a26b-6b400ceb3e8d.png)

**Math**
Here I have an image of manually doing the math for back propagation as I understand it. I wanted to know if this looks correct or if anything might be wrong as I have had a little trouble implementing it in code. I have gotten most of my information from this video [here](https://www.youtube.com/watch?v=tIeHLnjs5U8&t)

**Training**
From my understanding of training the network we feed forward an example then we preform back propagation. Back propagation is done by doing the math above. My clarification is what we do when we have these new weights. My current understand is once we have the new weights we multiply them by a learning rate and sum that to a matrix of the previous weights. This would be changing the weights by whatever each example wants to change them by. Here is a simple implementation: 
 
~~~python
weights_3 = np.random.rand(3,4)
weights_3_sum = weights_3.copy()
h2 = np.random.rand(3,1)

# derivative of cost function with respect to the weights (this will give the changes to the weights)
# output is output activations
# h2 is the second hidden layer,
# y is a vector that holds the the desired output
dc_dw3 = 2 (output - y) * self.sigmoid_derivative(output) * h2


# add the changes to weights
weights_3_sum += (0.01 * dc_dw3)

# continue this for every example

~~~

Any correction or piece of advice would be great.

**Comment by camoverride** [comment]
@GaelGil - define the variables _y_ and _output_ in your code so that I can run it. I can make assumptions about that these probably are, but it's easier if you define them so we can be on the same page

at a glance, your math looks correct (if you want me to critique it line-by-line, I suggest converting it to latex because it's easier to read that way).

However, consider the fact that all the variables you have listed - _b_, _h_, etc - refer to mathematical objects like numbers, matrices, etc. This kind of notation is helpful, but it's abstract. I suggest also following a less abstract (more concrete) approach by literally writing out these quantities (the numbers, matrices, etc) -- do the multiplication/addition by hand and show how the weights change during learning - you might need to draw several copies of the network during different learning steps. Of course, you won't be able to do this if you have a complicated network, but if you restrict yourself to a few nodes that should be manageable. This will let you "prove to yourself" that the abstract formulas you've written down actually work

Good job! multivariable calculus is hard and I didn't even learn this stuff until my early 20's

**Comment by GaelGil** [comment]
@camoverride I have worked on putting the math I wrote into latex. That can be found [here](https://github.com/GaelGil/notebooks/blob/master/back-propagation/back_propagatio_math.ipynb). 
I think I now realize what I have been doing wrong so some of the math may be wrong.

**Comment by camoverride** [comment]
Your math looks good! It's possible that there's a typo somewhere, but I'm not noticing any large errors. Can you also plug in some specific values and check the results for those values to see if it makes sense? You might need to use an even simpler network.

Also, see if you can get this working in code. You will probably want to use a library to calculate the error.

After you do that, check out the pytorch library, which is designed to keep track of gradients when doing matrix operations (lots of deep learning libraries are built on top of torch). A good assignment would be for you to read through [this tutorial](https://pytorch.org/tutorials/beginner/basics/tensorqs_tutorial.html) and come prepared to explain it to me along with examples.

**Comment by GaelGil** [comment]
> Your math looks good! It's possible that there's a typo somewhere, but I'm not noticing any large errors. Can you also plug in some specific values and check the results for those values to see if it makes sense? You might need to use an even simpler network.

Ill work on adding some samples.

**Comment by GaelGil** [comment]
@camoverride 
> Also, see if you can get this working in code. You will probably want to use a library to calculate the error.

I have been having some trouble understanding the back propagation. [Here](https://www.kaggle.com/wwsalmon/simple-mnist-nn-from-scratch-numpy-no-tf-keras/notebook) is an example that I'm going off of (amongst others). 


I know that we have to pass the activations  through some non linear function (sigmod, relu). But in the **forward propagation** section in this notebook (the last two lines) they pass the second layer activations z^2 into the softmax function. Then the output of that is the output layer as well. From my understanding those activations would still need to be multiplied by some weights and add some biases to get to the output. Is there something I'm missing here?

**Comment by GaelGil** [comment]
Another thing I am not really understanding is in the math in the section below that in the **back propagation**  section (the last three lines). As I am writing this I think I actually now understand it but the notation was a bit confusing. I can explain better in our next meeting.

**Comment by GaelGil** [comment]
I am closing this issue since I finished the math
